### PR TITLE
DAOS-3916 container: add number of handles to rdb, query output

### PR DESCRIPTION
**Pull request by kccain**
Before this change, the daos_cont_info_t structure returned by daos_cont_open() and daos_cont_query() includes an unnecessary value, ci_redun_fac (it is otherwise available in the properties). Additionally, callers to these APIs and daos_pool_filter_containers() have no visibility into the number of open handles on a container.

With this change, the daos_cont_info_t structure replaces the uint32_t member ci_redun_fac with ci_nhandles. The daos utility output (both human-readable and JSON format) is updated to output this quantity. The libdaos API minor version is also updated. The container open and query CaRT RPC protocols are  updated to return (where previously a pad value was specified) this 32-bit quantity. The protocol version is not changed, since it was already updated for this feature (metadata time tracking, and filtering based on metadata criteria). The daos_pool_filter_containers() API and implementation is updated to allow filtering based on a container's number of open handles. The daos_test pool and container test suites are updated to utilize the changes in the API functions and returned information.

The daos_engine container open, query and close handling is updated to get and increment/decrement the number of handles maintained in rdb. And engine code that performs pool/container metadata layout upgrades is updated to initialize the number of handles to zero when upgrading an older container (e.g., one created with DAOS v2.2). The pool filter containers handling is updated to process filters that include conditional tests based on the number of container open handles.

Other changes include:
- a small fix to pool_cont_filter_is_valid() engine code to inspect the key and comparison function specified in the RPC against the proper minimum and maximum bounds.
- a small change to the pool/container upgrade code for metadata times, updating modify time only (not open + modify times) for containers that already have metadata times in their rdb. This will fix a future upgrade scenario (from DAOS v2.4 to the following version).

Required-githooks: true

Signed-off-by: Kenneth Cain <kenneth.c.cain@intel.com>

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by kccain** [comment]
Build 4 is in progress (all except functional hw medium and large have run, no test failures so far). However, the master on which this is based has some known test failures that I expect to see in that medium/large hw testing. While that is getting sorted out independently, in parallel I would like to open this up for code reviews.

File: src/control/cmd/daos/container.go
Comment (mchaarawi) [inline src/control/cmd/daos/container.go:736]: as i understand from this PR, you are changing the output of daos cont query to not include:
`Container redundancy factor : x`
and have instead
`Number of open handles   : x`
?

the RF is something useful really that users would like to see on cont query. im ok with the change to the info structure to remove that since we can grab that programmatically from the props, but visually on cont query it has been helpful to have the RF there..

File: src/include/daos_types.h
Comment (mchaarawi) [inline src/include/daos_types.h:136]: why can't we have both (adding ci_nhandles at the end)?

File: src/tests/suite/daos_container.c
Comment (mchaarawi) [inline src/tests/suite/daos_container.c:2273]: we have internal functions:
dc_cont_hdl2redunfac and dc_cont_hdl2redunlvl
does it make sense to make a public API for those? (not necessarily with this PR)

File: src/tests/suite/daos_container.c
Comment (mchaarawi) [inline src/tests/suite/daos_container.c:57]: not quite sure i get why this should be 0 and not 1. we just opened 1 handle.

File: src/tests/suite/daos_container.c
Comment (mchaarawi) [inline src/tests/suite/daos_container.c:65]: same here, since we have 2 handles now, it should be 2?

File: src/tests/suite/daos_container.c
Comment (mchaarawi) [inline src/tests/suite/daos_container.c:78]: this one sounds correct because one handle is open.. so not sure i get it

File: src/include/daos_types.h
Comment (kccain) [reply src/include/daos_types.h:136]: It would break the ABI that to my understanding cannot be done until the next major API version increment (that cannot be done in DAOS 2.4).

File: src/tests/suite/daos_container.c
Comment (kccain) [reply src/tests/suite/daos_container.c:57]: Like for the fetching of the container metadata (open, modify) times I decided that the number of open handles returned should reflect the state before the open, instead of including the effect of the open.

File: src/tests/suite/daos_container.c
Comment (kccain) [reply src/tests/suite/daos_container.c:78]: Here this is  a container query (as opposed to the open whose "post" execution effect would be to increment the number of open handles). So in this case query returns the number that reflects the execution of the previous open.

File: src/include/daos_types.h
Comment (mchaarawi) [reply src/include/daos_types.h:136]: you are already breaking the API, which IMO is actually worse

File: src/tests/suite/daos_container.c
Comment (mchaarawi) [reply src/tests/suite/daos_container.c:57]: this is counter intuitive IMO. idk about the time values, but for this particular count, for a user we really should return the actual count of open handles at that point where they are doing the checking. the user has just opened the container. getting back 0 open handle count sounds wrong IMO.

**Review by mchaarawi — CHANGES_REQUESTED** [review]
we should probably add some more robust testing part of this PR:
1) verify open count from several ranks / threads opening the container
2) verify the behavior with l2g/g2l functions

File: src/include/daos_types.h
Comment (kccain) [reply src/include/daos_types.h:136]: Other than the daos_test are there other users of the ci_redun_fac structure member? Also in a previous commit I removed the other ci_redun_lvl member and we discussed the reasoning for making this subsequent change to ci_redun_fac as well https://github.com/daos-stack/daos/pull/9985#discussion_r952610355

This patch is a "nice to have" addition to what I've already implemented for daos_pool_filter_containers(). So, if we think that the API change (removing ci_redun_fac) is too severe, we can always just decide to abandon this patch (and give up the ability to track number of open handles). I think that would be a bad decision, since it seems this could be useful for a datamover use case (e.g., filter a pool's containers based on this and other criteria, to decide what data to migrate).

Comment (mchaarawi) [reply src/include/daos_types.h:136]: "Other than the daos_test are there other users of the ci_redun_fac structure member?"
i don't know.. it's in a public header, so anyone can be using it. in the middleware i control / support, it's not used directly.
my comment is just in relation to your comment: "It would break the ABI"
you are already doing worse than that by breaking the API :-) so my point is that just adding the variables is not going to be worse that replacing an existing one.

Comment (mchaarawi) [reply src/include/daos_types.h:136]: just to be clear, my change request is not for this. im fine if you want to just replace it here. hopefully no one is using it..
The main issue for this PR for me is the reported count in the num handles and not reporting the RF in the daos cont query (which seems unnecessary to remove in this case since you can just retrieve that from the open handle?)

File: src/tests/suite/daos_container.c
Comment (kccain) [reply src/tests/suite/daos_container.c:57]: Yes it is counter-intuitive. I'm going to look at returning nhandles to reflect the effect of the current open being executed.

And while doing this I will review the container metadata open time and whether it can be pre-updated to be consistent with what will be done with nhandles here. Open time has one tricky special case to account for (no update at all) - perhaps it is that case that made it seemingly difficult to pre-update the open time (when we do want the update to occur).

File: src/tests/suite/daos_container.c
Comment (kccain) [reply src/tests/suite/daos_container.c:78]: will pre-increment nhandles in the next push

File: src/tests/suite/daos_container.c
Comment (kccain) [reply src/tests/suite/daos_container.c:65]: yes, will fix in the next push

File: src/tests/suite/daos_container.c
Comment (kccain) [reply src/tests/suite/daos_container.c:57]: I have local commits that both pre-increment nhandles, and  pre-update the open time - will have both in the next push

File: src/container/srv_container.c
Comment (liw) [inline src/container/srv_container.c:2051]: [Not requesting changes] It looks like we don't need the special "PRE_<OP>" concepts or doing updates in functions named "get_<...>". What would the problem be if we simply write:

    lookup(tx, &nhandles)
    nhandles++ /* increment the variable for preforming the update call */
    update(tx, &nhandles)
    commit(tx)
    if tx cannot commit
        if necessary, do not publish nhandles and decrement the variable if needed

I think we may not even need the last `if` statement for cont_open.

**Review by liw — DISMISSED** [review]

File: src/container/srv_container.c
Comment (kccain) [reply src/container/srv_container.c:2051]: I was trying to consolidate the logic for looking up and updating this property across multiple calling locations, rather than copy/paste the code in each location. Agree that the word "get_" in the function can be misleading the way it has been implemented here.

**Comment by kccain** [comment]
> Test stage Functional Hardware Medium completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-10793/7/execution/node/1084/log

That is an instance of intermittent failure https://daosio.atlassian.net/browse/DAOS-12054

Running Functional HW Medium tests again (another build 8 triggered).

**Review by jolivier23 — COMMENTED** [review]
SCons change looks fine to me.   Though I think we may have already bumped it for 2.4 so it really only affects internal usage.

**Comment by kccain** [comment]
Addressed file conflict due to the landing of PR https://github.com/daos-stack/daos/pull/10746, and have pushed for CI testing.
Adding @kjacque to the review as well.

File: src/include/daos_types.h
Comment (kjacque) [reply src/include/daos_types.h:136]: I do find it a bit odd to break the API on a minor API version change. I don't feel strongly enough to block on it, but I do hope it doesn't come back to bite us.

**Review by kjacque — APPROVED** [review]

**Review by mchaarawi — DISMISSED** [review]
lgtm. although i prefer you get a +1 from @liw for the server side changes before you request landing as i'm not too familiar with that part.

File: src/include/daos_types.h
Comment (kjacque) [reply src/include/daos_types.h:136]: Oh, if it was added recently, e.g. after branching 2.2, that's fine. I thought it was already in a release, which could mean users expect it and we shouldn't remove it. But if it's only in master, in active development, I think we don't have to worry about it.

**Review by liw — APPROVED** [review]

File: src/container/srv_container.c
Comment (kccain) [inline src/container/srv_container.c:1974]: The type of out_v7 pointer variable declared here for CONT_OPEN_BYLABEL should have been struct cont_open_bylabel_v7_out *

This was the cause of the single test failure in build 9 - thought I had tested locally before pushing, but was mistaken. Fixing in a next/final push.

**Review by liw — APPROVED** [review]

**Comment by kccain** [comment]
Build 12 passed all tests that were run. However, there was a site reliability issue on functional hardware large, and medium clusters preventing those tests from running. I added a note about it in https://daosio.atlassian.net/browse/SRE-372.

Will try to retrigger functional hardware testing.

**Comment by kccain** [comment]
All tests passed, calling for final reviews

**Review by liw — DISMISSED** [review]

**Review by mjmac — COMMENTED** [review]
Go changes look fine to me. Did this ever get a test run with `Features: container` ? If not, my only concern would be that there might be some daily/weekly tests that are broken by the new container query output. I would hope not, but you know how that goes...

**Comment by kccain** [comment]
> Go changes look fine to me. Did this ever get a test run with `Features: container` ? If not, my only concern would be that there might be some daily/weekly tests that are broken by the new container query output. I would hope not, but you know how that goes...

it has not. I'm making an attempt to do so in a separate PR with identical code content (so as to not disturb the review state of this PR, i.e., not wanting to have to keep asking the reviewers to repeatedly visit PR and give +1).
https://github.com/daos-stack/daos/pull/10888
that PR for some reason didn't select any tests when run with Quick-Functional: true and Features: container so I have some debugging to do to see why that happened.

**Comment by kccain** [comment]
Oh, the branch now has conflicts so maybe it doesn't matter. Sigh. Time to push and test again.

**Comment by kccain** [comment]
PR passed all (expanded, with Features: container) tests after merging latest master to address file conflicts.

As discussed with @mjmac , adding daos-gatekeeper now. This PR has been reviewed numerous times and been given +1 by 2 or more developers, with the most recent changes to the patch being only for merging master & fixing file conflicts, and running expanded testing.

**Review by liw — APPROVED** [review]

**Pull request merged**
